DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/Transpose.hpp

```diff
@@ -5,65 +5,195 @@
+// 'k_pipe_width' == 1 and this cannot be done with a function.
+template <typename T, size_t k_num_cols_in, size_t k_pipe_width,
+          typename MatrixInPipe, typename MatrixOutPipe>
+struct Transposer {
+  void operator()() const {
+    using PipeType = NTuple<T, k_pipe_width>;
+
+    // This is a scratch pad memory that we will use to do the transpose.
+    // We read the data in from a pipe (k_pipe_width elements at at time),
+    // store it in this memory in row-major format and read it out in
+    // column-major format (again, k_pipe_width elements at a time).
+    constexpr int kNumScratchMemCopies = 4;
+    constexpr unsigned char kNumScratchMemCopiesBitMask = 0x03;
+    constexpr int kBankwidth = k_pipe_width * sizeof(T);
+    // NO-FORMAT comments are for clang-format
+    [[intel::numbanks(1)]]                   // NO-FORMAT: Attribute
+    [[intel::bankwidth(kBankwidth)]]         // NO-FORMAT: Attribute
+    [[intel::private_copies(1)]]             // NO-FORMAT: Attribute
+    [[intel::max_replicates(k_pipe_width)]]  // NO-FORMAT: Attribute
+    T scratch[kNumScratchMemCopies][k_pipe_width][k_num_cols_in];
+
+    // track the status of each of the buffers
+    // NO-FORMAT comments are for clang-format
+    [[intel::fpga_register]]  // NO-FORMAT: Attribute
+    bool ready_to_send[kNumScratchMemCopies] = {false, false, false, false};
+
+    unsigned char rx_buffer = 0;
+    unsigned short rx_count = 0;
+    unsigned char tx_buffer = 0;
+    unsigned char tx_col = 0;
+    bool last_tx_col = false;
+
+    // create a 'pipeline' for the almost full signal
+    constexpr int kAlmostFullPipeDepth = 2;
+    NTuple<bool, kAlmostFullPipeDepth> almost_full_pipe;
+    UnrolledLoop<kAlmostFullPipeDepth>([&](auto pipe_stage) {
+      almost_full_pipe.template get<pipe_stage>() = false;
+    });
+
+    // NO-FORMAT comments are for clang-format
+    [[intel::ii(1)]]  // NO-FORMAT: Attribute
+    while (1) {
+      // capture current value of all status variables as we begin each loop
+      // iteration
+      unsigned char cur_rx_buffer = rx_buffer;
+      unsigned short cur_rx_count = rx_count;
+      unsigned char cur_tx_buffer = tx_buffer;
+      unsigned char cur_tx_col = tx_col;
+      bool cur_last_tx_col = last_tx_col;
+
+      // Create a 'pipeline' for the almost full signal.
+      // Calculate almost full at the end of the pipeline, and on each loop
+      // iteration we shift the data down the pipeline.  Since we are using
+      // an 'almost' full signal, we don't need the result right away, we
+      // can wait several loop iterations.  This allows us to break
+      // dependencies between loop iterations and improve FMAX.
+      UnrolledLoop<kAlmostFullPipeDepth - 1>([&](auto pipe_stage) {
+        almost_full_pipe.template get<pipe_stage>() =
+            almost_full_pipe.template get<pipe_stage + 1>();
+      });
+      bool cur_almost_full = almost_full_pipe.template get<0>();
```

Comment:
Used first() and last() here and in InputDemux (FYI the template keyword had to be removed, since first and last are not templated functions).

After the loop was intentional.  If I put it before the UnrolledLoop, then I've always got effectively a minimum of 2 pipeline stages.  This pattern allows the developer to easily experiment with no pipeline stages without having to rip all the code out.